ckan/migration/migrate_package_activity.py

```diff
@@ -0,0 +1,263 @@
+    print_errors(errors)
+
+
+def migrate_dataset(dataset_name, errors):
+    # monkey patch the legacy versions of code back into CKAN - so it has the
+    # revision functionality needed for this migration
+    import ckan.lib.dictization.model_dictize as model_dictize
+    try:
+        import ckan.migration.revision_legacy_code as revision_legacy_code
+    except ImportError:
+        # convenient to look for it in the current directory if you just
+        # download these files because you are upgrading an older ckan
+        import revision_legacy_code
+    model_dictize.package_dictize = \
+        revision_legacy_code.package_dictize_with_revisions
+
+    import ckan.logic as logic
+    from ckan import model
+
+    context = get_context()
+    # 'hidden' activity is that by site_user, such as harvests, which are
+    # not shown in the activity stream because they can be too numerous.
+    # However these do have Activity objects, and if a hidden Activity is
+    # followed be a non-hidden one and you look at the changes of that
+    # non-hidden Activity, then it does a diff with the hidden one (rather than
+    # the most recent non-hidden one), so it is important to store the
+    # package_dict in hidden Activity objects.
+    context[u'include_hidden_activity'] = True
+    package_activity_stream = logic.get_action(u'package_activity_list')(
+        context, {u'id': dataset_name})
+    num_activities = len(package_activity_stream)
+    if not num_activities:
+        print(u'  No activities')
+
+    context[u'for_view'] = False
+    # Iterate over this package's existing activity stream objects
+    for i, activity in enumerate(package_activity_stream):
+        # e.g. activity =
+        # {'activity_type': u'changed package',
+        #  'id': u'62107f87-7de0-4d17-9c30-90cbffc1b296',
+        #  'object_id': u'7c6314f5-c70b-4911-8519-58dc39a8e340',
+        #  'revision_id': u'c3e8670a-f661-40f4-9423-b011c6a3a11d',
+        #  'timestamp': '2018-04-20T16:11:45.363097',
+        #  'user_id': u'724273ac-a5dc-482e-add4-adaf1871f8cb'}
+        print(u'  activity {}/{} {}'.format(
+              i + 1, num_activities, activity[u'timestamp']))
+
+        # we need activity.data and using the ORM is the fastest
+        activity_obj = model.Session.query(model.Activity).get(activity[u'id'])
+        if u'resources' in activity_obj.data.get(u'package', {}):
+            print(u'    activity has full dataset already recorded'
+                  ' - no action')
+            continue
+
+        # get the dataset as it was at this revision
+        context[u'revision_id'] = activity[u'revision_id']
+        # call package_show just as we do in package.py:activity_stream_item(),
+        # only with a revision_id
+        try:
+            dataset = logic.get_action(u'package_show')(
+                context,
+                {u'id': activity[u'object_id'], u'include_tracking': False})
+        except logic.NotFound as exc:
```

Comment:
other errors are possible here, an old revision might have data that doesn't conform to the latest dataset schema or is incompatible with currently installed plugins. Those errors shouldn't prevent the migration from going ahead.

Reply:
I've put a fix in https://github.com/ckan/ckan/commit/8f30186f8